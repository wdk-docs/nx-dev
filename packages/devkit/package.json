{
  "name": "@nrwl/devkit",
  "version": "0.0.1",
  "description": "Smart, Fast and Extensible Build System",
  "repository": {
    "type": "git",
    "url": "https://github.com/nrwl/nx.git",
    "directory": "packages/devkit"
  },
  "keywords": [
    "Monorepo",
    "Angular",
    "React",
    "Web",
    "Node",
    "Nest",
    "Jest",
    "Cypress",
    "CLI"
  ],
  "main": "./index.js",
  "typings": "./index.d.ts",
  "author": "Victor Savkin",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/nrwl/nx/issues"
  },
  "homepage": "https://nx.dev",
  "dependencies": {
    "@nrwl/tao": "*",
    "ejs": "^3.1.5",
    "ignore": "^5.0.4",
    "rxjs": "^6.5.4",
    "semver": "7.3.4",
    "tslib": "^2.3.0"
  }
}
