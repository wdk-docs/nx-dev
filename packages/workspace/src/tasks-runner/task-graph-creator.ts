import type {
  ProjectGraph,
  TargetDependencyConfig,
  Task,
  TaskGraph,
} from '@nrwl/devkit';
import { getDependencyConfigs } from './utils';

export class TaskGraphCreator {
  constructor(
    private readonly projectGraph: ProjectGraph,
    private readonly defaultTargetDependencies: Record<
      string,
      TargetDependencyConfig[]
    >
  ) {}

  createTaskGraph(tasks: Task[]): TaskGraph {
    const graph: TaskGraph = {
      roots: [],
      tasks: {},
      dependencies: {},
    };
    for (const task of tasks) {
      this.addTaskToGraph(task, graph);

      const dependencyConfigs = getDependencyConfigs(
        task.target,
        this.defaultTargetDependencies,
        this.projectGraph
      );

      if (!dependencyConfigs) {
        continue;
      }

      this.addTaskDependencies(task, dependencyConfigs, tasks, graph);
    }

    graph.roots = Object.keys(graph.dependencies).filter(
      (k) => graph.dependencies[k].length === 0
    );

    return graph;
  }

  private addTaskDependencies(
    task: Task,
    dependencyConfigs: TargetDependencyConfig[],
    tasks: Task[],
    graph: TaskGraph
  ) {
    for (const dependencyConfig of dependencyConfigs) {
      if (dependencyConfig.projects === 'self') {
        for (const t of tasks) {
          if (
            t.target.project === task.target.project &&
            t.target.target === dependencyConfig.target
          ) {
            graph.dependencies[task.id].push(t.id);
          }
        }
      } else if (dependencyConfig.projects === 'dependencies') {
        const seen = new Set<string>();

        this.addDependencies(
          task.target.project,
          dependencyConfig.target,
          tasks,
          graph,
          task.id,
          seen
        );
      }
    }
  }

  private addDependencies(
    project: string,
    target: string,
    tasks: Task[],
    graph: TaskGraph,
    taskId: string,
    seen: Set<string>
  ) {
    seen.add(project);
    const dependencies = this.projectGraph.dependencies[project];
    if (dependencies) {
      const projectDependencies = dependencies.map(
        (dependency) => dependency.target
      );
      for (const projectDependency of projectDependencies) {
        if (seen.has(projectDependency)) {
          continue;
        }
        const dependency = this.findTask(
          { project: projectDependency, target },
          tasks
        );
        if (dependency) {
          graph.dependencies[taskId].push(dependency.id);
        } else {
          this.addDependencies(
            projectDependency,
            target,
            tasks,
            graph,
            taskId,
            seen
          );
        }
      }
    }
  }

  private findTask(
    { project, target }: { project: string; target: string },
    tasks: Task[]
  ): Task {
    return tasks.find(
      (t) => t.target.project === project && t.target.target === target
    );
  }

  private addTaskToGraph(task: Task, graph: TaskGraph) {
    graph.tasks[task.id] = task;
    graph.dependencies[task.id] = [];
  }
}
