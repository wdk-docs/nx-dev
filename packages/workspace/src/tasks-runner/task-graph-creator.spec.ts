import { createTasksForProjectToRun } from './run-command';
import { DependencyType, ProjectGraph } from '@nrwl/devkit';
import { TaskGraphCreator } from './task-graph-creator';

describe('TaskGraphCreator', () => {
  it('should return empty for an empty array', () => {
    const empty = new TaskGraphCreator(
      {
        nodes: {},
        dependencies: {},
      },
      {}
    ).createTaskGraph([]);
    expect(empty).toEqual({
      roots: [],
      tasks: {},
      dependencies: {},
    });
  });

  it('should order tasks based on project dependencies', () => {
    const taskGraph = new TaskGraphCreator(
      {
        nodes: {
          child1: { type: 'lib' },
          child2: { type: 'lib' },
          parent: { type: 'lib' },
          grandparent: { type: 'lib' },
        } as any,
        dependencies: {
          child1: [],
          child2: [],
          parent: [
            {
              source: 'parent',
              target: 'child1',
              type: DependencyType.static,
            },
            {
              source: 'parent',
              target: 'child2',
              type: DependencyType.static,
            },
          ],
          grandparent: [
            {
              source: 'grandparent',
              target: 'parent',
              type: DependencyType.static,
            },
          ],
        },
      },
      {
        build: [
          {
            target: 'build',
            projects: 'dependencies',
          },
        ],
      }
    ).createTaskGraph([
      {
        id: 'parent:build',
        target: { project: 'parent', target: 'build' },
      },
      {
        id: 'child1:build',
        target: { project: 'child1', target: 'build' },
      },
      {
        id: 'child2:build',
        target: { project: 'child2', target: 'build' },
      },
      {
        id: 'grandparent:build',
        target: { project: 'grandparent', target: 'build' },
      },
    ] as any);

    expect(taskGraph.roots).toEqual(['child1:build', 'child2:build']);

    expect(Object.keys(taskGraph.tasks)).toEqual([
      'parent:build',
      'child1:build',
      'child2:build',
      'grandparent:build',
    ]);

    expect(taskGraph.dependencies).toEqual({
      'child1:build': [],
      'child2:build': [],
      'grandparent:build': ['parent:build'],
      'parent:build': ['child1:build', 'child2:build'],
    });
  });

  it('should support custom targets that require strict ordering', () => {
    const taskGraph = new TaskGraphCreator(
      {
        nodes: {
          child1: { type: 'lib', data: { targets: { custom: {} } } },
          parent: { type: 'lib', data: { targets: { custom: {} } } },
        } as any,
        dependencies: {
          child1: [],
          parent: [
            {
              source: 'parent',
              target: 'child1',
              type: DependencyType.static,
            },
          ],
        },
      },
      {
        custom: [
          {
            target: 'custom',
            projects: 'dependencies',
          },
        ],
      }
    ).createTaskGraph([
      {
        id: 'parent:custom',
        target: { project: 'parent', target: 'custom' },
      },
      {
        id: 'child1:custom',
        target: { project: 'child1', target: 'custom' },
      },
    ] as any);

    expect(taskGraph.roots).toEqual(['child1:custom']);

    expect(Object.keys(taskGraph.tasks)).toEqual([
      'parent:custom',
      'child1:custom',
    ]);

    expect(taskGraph.dependencies).toEqual({
      'child1:custom': [],
      'parent:custom': ['child1:custom'],
    });

    const noDeps = new TaskGraphCreator(
      {
        nodes: {
          child1: { type: 'lib', data: { targets: { custom: {} } } },
          parent: { type: 'lib', data: { targets: { custom: {} } } },
        } as any,
        dependencies: {
          child1: [],
          parent: [
            {
              source: 'parent',
              target: 'child1',
              type: DependencyType.static,
            },
          ],
        },
      },
      {}
    ).createTaskGraph([
      {
        id: 'parent:custom',
        target: { project: 'parent', target: 'custom' },
      },
      {
        id: 'child1:custom',
        target: { project: 'child1', target: 'custom' },
      },
    ] as any);

    expect(noDeps.roots).toEqual(['parent:custom', 'child1:custom']);
  });

  describe('(tasks with dependency configurations)', () => {
    let projectGraph: ProjectGraph;
    beforeEach(() => {
      projectGraph = {
        nodes: {
          app1: {
            type: 'app',
            name: 'app1',
            data: {
              targets: {
                build: {
                  dependsOn: [
                    {
                      target: 'build',
                      projects: 'dependencies',
                    },
                  ],
                },
              },
              root: 'app1-root',
              files: [],
            },
          },
          app2: {
            type: 'app',
            name: 'app2',
            data: {
              targets: {
                build: {
                  dependsOn: [
                    {
                      target: 'build',
                      projects: 'dependencies',
                    },
                  ],
                },
              },
              root: 'app2-root',
              files: [],
            },
          },
          common1: {
            type: 'lib',
            name: 'common1',
            data: {
              targets: {
                build: {
                  dependsOn: [
                    {
                      target: 'build',
                      projects: 'dependencies',
                    },
                  ],
                },
              },
              root: 'common1-root',
              files: [],
            },
          },
          common2: {
            type: 'lib',
            name: 'common2',
            data: {
              targets: {
                build: {
                  dependsOn: [
                    {
                      target: 'build',
                      projects: 'dependencies',
                    },
                  ],
                },
              },
              root: 'common2-root',
              files: [],
            },
          },
        },
        dependencies: {
          app1: [
            {
              source: 'app1',
              target: 'common1',
              type: DependencyType.static,
            },
          ],
          app2: [
            {
              source: 'app2',
              target: 'common2',
              type: DependencyType.static,
            },
          ],
          common1: [],
          common2: [],
        },
      };
    });

    it('should create task graph (builds depend on build of dependencies)', () => {
      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1, projectGraph.nodes.app2],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );

      const taskGraph = new TaskGraphCreator(projectGraph, {}).createTaskGraph(
        tasks
      );

      expect(taskGraph).toMatchSnapshot();
    });

    it('should create a task graph (builds depend on builds of dependencies even with intermediate projects)', () => {
      delete projectGraph.nodes.common1.data.targets.build;
      projectGraph.dependencies.common1.push({
        type: DependencyType.static,
        source: 'common1',
        target: 'common2',
      });

      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );

      const taskGraph = new TaskGraphCreator(projectGraph, {}).createTaskGraph(
        tasks
      );

      expect(taskGraph).toMatchSnapshot();
    });

    it('should create a task graph (builds depend on builds of dependencies with intermediate projects and circular dependencies between projects)', () => {
      delete projectGraph.nodes.common1.data.targets.build;
      projectGraph.dependencies.common1.push({
        type: DependencyType.static,
        source: 'common1',
        target: 'common2',
      });

      projectGraph.dependencies.common2.push({
        type: DependencyType.static,
        source: 'common2',
        target: 'common1',
      });

      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );

      const taskGraph = new TaskGraphCreator(projectGraph, {}).createTaskGraph(
        tasks
      );

      expect(taskGraph).toMatchSnapshot();
    });

    it('should create a task graph (builds depend on builds of dependencies with intermediate projects and circular dependencies between projects) 2', () => {
      delete projectGraph.nodes.common1.data.targets.build;
      projectGraph.dependencies.common1.push({
        type: DependencyType.static,
        source: 'common1',
        target: 'common2',
      });

      delete projectGraph.nodes.common2.data.targets.build;
      projectGraph.dependencies.common2.push({
        type: DependencyType.static,
        source: 'common2',
        target: 'common3',
      });

      projectGraph.nodes.common3 = {
        name: 'common3',
        type: 'lib',
        data: {
          root: 'common3',
          targets: {
            build: {},
          },
        },
      };
      projectGraph.dependencies.common3 = [];

      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );

      const taskGraph = new TaskGraphCreator(projectGraph, {}).createTaskGraph(
        tasks
      );

      expect(taskGraph).toMatchSnapshot();
    });

    it('should create task graph (builds depend on build of dependencies and prebuild of self)', () => {
      projectGraph.nodes.app1.data.targets.prebuild = {};
      projectGraph.nodes.app2.data.targets.prebuild = {};
      projectGraph.nodes.app1.data.targets.build.dependsOn.push({
        projects: 'self',
        target: 'prebuild',
      });
      projectGraph.nodes.app2.data.targets.build.dependsOn.push({
        projects: 'self',
        target: 'prebuild',
      });
      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1, projectGraph.nodes.app2],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );
      const taskGraph = new TaskGraphCreator(projectGraph, {}).createTaskGraph(
        tasks
      );

      expect(taskGraph).toMatchSnapshot();
    });

    it('should create task graph (builds depend on build of dependencies, builds depend on prebuilds)', () => {
      projectGraph.nodes.app1.data.targets.prebuild = {
        dependsOn: [
          {
            target: 'build',
            projects: 'dependencies',
          },
        ],
      };
      projectGraph.nodes.app2.data.targets.prebuild = {
        dependsOn: [
          {
            target: 'build',
            projects: 'dependencies',
          },
        ],
      };
      projectGraph.nodes.app1.data.targets.build.dependsOn.push({
        projects: 'self',
        target: 'prebuild',
      });
      projectGraph.nodes.app2.data.targets.build.dependsOn.push({
        projects: 'self',
        target: 'prebuild',
      });
      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1, projectGraph.nodes.app2],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );
      const taskGraph = new TaskGraphCreator(projectGraph, {}).createTaskGraph(
        tasks
      );

      expect(taskGraph).toMatchSnapshot();
    });

    it('should create task graph (builds depend on build of dependencies, builds depend on prebuilds)', () => {
      projectGraph.nodes.common1.data.targets =
        projectGraph.nodes.common2.data.targets = {
          prebuild: {},
          build: {
            dependsOn: [
              {
                target: 'build',
                projects: 'dependencies',
              },
              {
                target: 'prebuild',
                projects: 'self',
              },
            ],
          },
        };
      projectGraph.nodes.app1.data.targets =
        projectGraph.nodes.app2.data.targets = {
          build: {
            dependsOn: [
              {
                target: 'build',
                projects: 'dependencies',
              },
            ],
          },
        };
      const tasks = createTasksForProjectToRun(
        [projectGraph.nodes.app1, projectGraph.nodes.app2],
        {
          target: 'build',
          configuration: undefined,
          overrides: {},
        },
        projectGraph,
        null
      );

      const taskGraph = new TaskGraphCreator(projectGraph, {}).createTaskGraph(
        tasks
      );

      expect(taskGraph).toMatchSnapshot();
    });
  });
});
