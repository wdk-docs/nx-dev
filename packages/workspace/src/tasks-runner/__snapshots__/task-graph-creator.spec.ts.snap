// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskGraphCreator (tasks with dependency configurations) should create a task graph (builds depend on builds of dependencies even with intermediate projects) 1`] = `
Object {
  "dependencies": Object {
    "app1:build": Array [
      "common2:build",
    ],
    "common2:build": Array [],
  },
  "roots": Array [
    "common2:build",
  ],
  "tasks": Object {
    "app1:build": Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    "common2:build": Object {
      "id": "common2:build",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "build",
      },
    },
  },
}
`;

exports[`TaskGraphCreator (tasks with dependency configurations) should create a task graph (builds depend on builds of dependencies with intermediate projects and circular dependencies between projects) 1`] = `
Object {
  "dependencies": Object {
    "app1:build": Array [
      "common2:build",
    ],
    "common2:build": Array [],
  },
  "roots": Array [
    "common2:build",
  ],
  "tasks": Object {
    "app1:build": Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    "common2:build": Object {
      "id": "common2:build",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "build",
      },
    },
  },
}
`;

exports[`TaskGraphCreator (tasks with dependency configurations) should create a task graph (builds depend on builds of dependencies with intermediate projects and circular dependencies between projects) 2 1`] = `
Object {
  "dependencies": Object {
    "app1:build": Array [
      "common3:build",
    ],
    "common3:build": Array [],
  },
  "roots": Array [
    "common3:build",
  ],
  "tasks": Object {
    "app1:build": Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    "common3:build": Object {
      "id": "common3:build",
      "overrides": Object {},
      "projectRoot": "common3",
      "target": Object {
        "configuration": undefined,
        "project": "common3",
        "target": "build",
      },
    },
  },
}
`;

exports[`TaskGraphCreator (tasks with dependency configurations) should create task graph (builds depend on build of dependencies and prebuild of self) 1`] = `
Object {
  "dependencies": Object {
    "app1:build": Array [
      "common1:build",
      "app1:prebuild",
    ],
    "app1:prebuild": Array [],
    "app2:build": Array [
      "common2:build",
      "app2:prebuild",
    ],
    "app2:prebuild": Array [],
    "common1:build": Array [],
    "common2:build": Array [],
  },
  "roots": Array [
    "common1:build",
    "app1:prebuild",
    "common2:build",
    "app2:prebuild",
  ],
  "tasks": Object {
    "app1:build": Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    "app1:prebuild": Object {
      "id": "app1:prebuild",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "prebuild",
      },
    },
    "app2:build": Object {
      "id": "app2:build",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "build",
      },
    },
    "app2:prebuild": Object {
      "id": "app2:prebuild",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "prebuild",
      },
    },
    "common1:build": Object {
      "id": "common1:build",
      "overrides": Object {},
      "projectRoot": "common1-root",
      "target": Object {
        "configuration": undefined,
        "project": "common1",
        "target": "build",
      },
    },
    "common2:build": Object {
      "id": "common2:build",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "build",
      },
    },
  },
}
`;

exports[`TaskGraphCreator (tasks with dependency configurations) should create task graph (builds depend on build of dependencies) 1`] = `
Object {
  "dependencies": Object {
    "app1:build": Array [
      "common1:build",
    ],
    "app2:build": Array [
      "common2:build",
    ],
    "common1:build": Array [],
    "common2:build": Array [],
  },
  "roots": Array [
    "common1:build",
    "common2:build",
  ],
  "tasks": Object {
    "app1:build": Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    "app2:build": Object {
      "id": "app2:build",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "build",
      },
    },
    "common1:build": Object {
      "id": "common1:build",
      "overrides": Object {},
      "projectRoot": "common1-root",
      "target": Object {
        "configuration": undefined,
        "project": "common1",
        "target": "build",
      },
    },
    "common2:build": Object {
      "id": "common2:build",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "build",
      },
    },
  },
}
`;

exports[`TaskGraphCreator (tasks with dependency configurations) should create task graph (builds depend on build of dependencies, builds depend on prebuilds) 1`] = `
Object {
  "dependencies": Object {
    "app1:build": Array [
      "common1:build",
      "app1:prebuild",
    ],
    "app1:prebuild": Array [
      "common1:build",
    ],
    "app2:build": Array [
      "common2:build",
      "app2:prebuild",
    ],
    "app2:prebuild": Array [
      "common2:build",
    ],
    "common1:build": Array [],
    "common2:build": Array [],
  },
  "roots": Array [
    "common1:build",
    "common2:build",
  ],
  "tasks": Object {
    "app1:build": Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    "app1:prebuild": Object {
      "id": "app1:prebuild",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "prebuild",
      },
    },
    "app2:build": Object {
      "id": "app2:build",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "build",
      },
    },
    "app2:prebuild": Object {
      "id": "app2:prebuild",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "prebuild",
      },
    },
    "common1:build": Object {
      "id": "common1:build",
      "overrides": Object {},
      "projectRoot": "common1-root",
      "target": Object {
        "configuration": undefined,
        "project": "common1",
        "target": "build",
      },
    },
    "common2:build": Object {
      "id": "common2:build",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "build",
      },
    },
  },
}
`;

exports[`TaskGraphCreator (tasks with dependency configurations) should create task graph (builds depend on build of dependencies, builds depend on prebuilds) 2`] = `
Object {
  "dependencies": Object {
    "app1:build": Array [
      "common1:build",
    ],
    "app2:build": Array [
      "common2:build",
    ],
    "common1:build": Array [
      "common1:prebuild",
    ],
    "common1:prebuild": Array [],
    "common2:build": Array [
      "common2:prebuild",
    ],
    "common2:prebuild": Array [],
  },
  "roots": Array [
    "common1:prebuild",
    "common2:prebuild",
  ],
  "tasks": Object {
    "app1:build": Object {
      "id": "app1:build",
      "overrides": Object {},
      "projectRoot": "app1-root",
      "target": Object {
        "configuration": undefined,
        "project": "app1",
        "target": "build",
      },
    },
    "app2:build": Object {
      "id": "app2:build",
      "overrides": Object {},
      "projectRoot": "app2-root",
      "target": Object {
        "configuration": undefined,
        "project": "app2",
        "target": "build",
      },
    },
    "common1:build": Object {
      "id": "common1:build",
      "overrides": Object {},
      "projectRoot": "common1-root",
      "target": Object {
        "configuration": undefined,
        "project": "common1",
        "target": "build",
      },
    },
    "common1:prebuild": Object {
      "id": "common1:prebuild",
      "overrides": Object {},
      "projectRoot": "common1-root",
      "target": Object {
        "configuration": undefined,
        "project": "common1",
        "target": "prebuild",
      },
    },
    "common2:build": Object {
      "id": "common2:build",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "build",
      },
    },
    "common2:prebuild": Object {
      "id": "common2:prebuild",
      "overrides": Object {},
      "projectRoot": "common2-root",
      "target": Object {
        "configuration": undefined,
        "project": "common2",
        "target": "prebuild",
      },
    },
  },
}
`;
